egs/librimix/ConvTasNet/eval.py

```diff
@@ -41,11 +53,101 @@
 )
+parser.add_argument(
+    "--compute-wer", type=int, default=0, help="Compute WER using ESPNet's pretrained model"
+)
+
+COMPUTE_METRICS = ["si_sdr", "sdr", "sir", "sar", "stoi"]
+ASR_MODEL_PATH = (
+    "Shinji Watanabe/librispeech_asr_train_asr_transformer_e18_raw_bpe_sp_valid.acc.best"
+)
+
+
+def update_compute_metrics(compute_wer, metric_list):
+    if not compute_wer:
+        return metric_list
+    try:
+        from espnet2.bin.asr_inference import Speech2Text
+        from espnet_model_zoo.downloader import ModelDownloader
+    except ModuleNotFoundError:
+        import warnings
+
+        warnings.warn("Couldn't find espnet installation. Continuing without.")
+        return metric_list
+    return metric_list + ["wer"]
+
+
+class MockTracker:
+    def __init__(self, *args, **kwargs):
+        pass
 
-compute_metrics = ["si_sdr", "sdr", "sir", "sar", "stoi"]
+    def __call__(self, *args, **kwargs):
+        return dict()
+
+
+class WERTracker:
+    def __init__(self, model_name, trans_df):
+        from espnet2.bin.asr_inference import Speech2Text
+        from espnet_model_zoo.downloader import ModelDownloader
+
+        self.model_name = model_name
+        d = ModelDownloader()
+        self.asr_model = Speech2Text(**d.download_and_unpack(model_name))
+        self.input_txt_list = []
+        self.output_txt_list = []
+        self.sample_rate = int(d.data_frame[d.data_frame["name"] == model_name]["fs"])
+        self.trans_df = trans_df
+        self.trans_dic = self._df_to_dict(trans_df)
+
+    def __call__(
+        self, *, mix: np.ndarray, est_sources: np.ndarray, sample_rate: int, wav_id: List[str]
+    ):
+        """Compute and store best hypothesis for the mixture and the estimates"""
+        if sample_rate != self.sample_rate:
+            mix, est_sources = self.resample(
+                mix, est_sources, fs_from=sample_rate, fs_to=self.sample_rate
+            )
+        input_wer = output_wer = 0.0
+        # Count the mixture output for each speaker
+        txt = self.predict_hypothesis(mix)
+        for tmp_id in wav_id:
+            input_wer += wer(truth=self.trans_dic[tmp_id], hypothesis=txt) / len(wav_id)
+            self.input_txt_list.append(dict(utt_id=tmp_id, text=txt))
+        # Average WER for the estimate pair
+        for est, tmp_id in zip(est_sources, wav_id):
+            txt = self.predict_hypothesis(est)
+            output_wer += wer(truth=self.trans_dic[tmp_id], hypothesis=txt) / len(wav_id)
+            self.output_txt_list.append(dict(utt_id=tmp_id, text=txt))
+        return dict(input_wer=input_wer, wer=output_wer)
+
+    def predict_hypothesis(self, wav):
+        nbests = self.asr_model(wav)
+        text, *_ = nbests[0]
+        return text
+
+    def compute_wer(self, transcriptions):
+        # Is average WER the average over individual WER?
+        # Or computed overall with all S, D, I?
+        # Is it different than the average from call (probably)
+        return
```

Comment:
Would love an answer to that

Reply:
https://github.com/kaldi-asr/kaldi/blob/85a3dd5f0b71e419abf1169a26b759bfc423a543/src/bin/compute-wer.cc#L94
This is the way to go: compute it over all the dataset (e.g. test set) by summing the errors of all utterances then normalize by the sum of n_words. If you compute it by utterance and then take average of WERs say you have on word in one sequence and you miss that you have there 100% WER there. In another you have 20 words and you get them all correct. If you take your average you get 50% which is not really reperesentative of the performance at this point.

Reply:
This is what I thought, thanks !
How expensive is the computation of WER? At first, I had planned to accumulate all the hypothesis/transcripts and compute the overall WER at the end. I still log all of it so we can compute the average WER in the end (supported by jiwer, which accounts for total number of words, as we want). 

BTW, even if less critical, we do this "mistake" on SI-SDR, 3s examples and 12s examples have the same weight on the averaged metric.

Reply:
> This is what I thought, thanks !
> How expensive is the computation of WER? At first, I had planned to accumulate all the hypothesis/transcripts and compute the overall WER at the end. I still log all of it so we can compute the average WER in the end (supported by jiwer, which accounts for total number of words, as we want).

It depends on the internal algorithm (o(N) or (o(N^2)) and I don't know the details of this library.
Some toolkit takes a very long time for the WER computation of the long sequence.
I think your idea would be fine.

Some people may want to know the breakdown numbers per utterance or per speaker, but this would be too much for now.
This is one reason that we still use NIST sclite, which provides such analysis information, but it is not a pythonic solution.
(Another reason for using NIST toolkit is that I want to stick to use a standardized toolkit for the evaluation metric. It is a disaster if the scoring toolkit includes some errors or some dialects.).

Also, I recommend you generate the number of total word, insertion, deletion, and substitution counts, as well as WER.
This is good debugging information.
I often encountered the issue that people don't use the correct test set.
If we display the number of sentences and words, we can easily detect such mistakes.
Also, if we display insertion, deletion, and substitution counts, we can easily detect what issue causes the error.
Speech enhancement or separation tends to increase deletion
(if the speech is over suppressed) and the insertion (if there are interference speakers).

Reply:
Thank you very much for your feedback!
We'll log the numbers (ins, del, sub, tot, WER) per utterance then, as is done for the other metrics.

Reply:
For reference: 
Opened a PR in jiwer to access those numbers: https://github.com/jitsi/jiwer/pull/35